Update to command line messages for tctl create commands

hegurhurgurk opened the pull request:
Updated the create commands to make their output more uniform. Also updated the upsert verb to use more accurate verbs.

hegurhurgurk committed:
Update Create commands in resource_command.go

updated tctl command line output for create commands to be more standardized

hegurhurgurk committed:
Update resource_command.go

zmb3 commented:
Thanks for the contribution, but I'm not convinced this extra logic is something we need at this point.

ibeckermayer commented:
> Thanks for the contribution, but I'm not convinced this extra logic is something we need at this point.

@zmb3 I put @hegurhurgurk up to this, the reason for the extra logic is to make it so that for all resources, if the resource already exists, users must call `tctl create -f resource.yaml` in order to overwrite them. 

Before these changes, some resources required `-f` to be overwritten, while others didn't, with no particular rhyme or reason.

zmb3 commented:
> Before these changes, some resources required -f to be overwritten, while others didn't, with no particular rhyme or reason.

Sure, but now we may break existing scripts or automation.

Additionally, calling get before an upsert as an attempt to detect whether the item was created or updated doesn't seem right. (It makes the operation take twice as long since there are now 2 round trips, and you can't be sure the resource wasn't inserted in between the get and the upsert.)

ibeckermayer commented:
> Sure, but now we may break existing scripts or automation.

In that case how about we avoid any backports and keep this as a change in a major release (breaking API changes).

> Additionally, calling get before an upsert as an attempt to detect whether the item was created or updated doesn't seem right. (It makes the operation take twice as long since there are now 2 round trips, and you can't be sure the resource wasn't inserted in between the get and the upsert.)

Valid point, the counterpoint being that this is how we're already handling this behavior for many of the resources in `tctl create`, e.g. for `tctl create user` and `tctl create role`.

https://github.com/gravitational/teleport/blob/9c55a33e0143f1e8bd6063872e139705a2e47941/tool/tctl/common/resource_command.go#L441

https://github.com/gravitational/teleport/blob/9c55a33e0143f1e8bd6063872e139705a2e47941/tool/tctl/common/resource_command.go#L388

Two round trips is obviously worse than one, but I think data integrity (making sure the user is being explicit by making them use `-f`) and API consistency (all else equal, that behavior is the same for all resources) is a more important goal than performance for these typically-one-off command line operations. 

I don't know if there's any way to solve for that potential race condition given our current system without adding some sort of distributed locking mechanism on resources.

Ultimately I think we should
1. Make sure all of our `CreateX` methods return `trace.AlreadyExists` if the resource already exists (relates to https://github.com/gravitational/teleport/issues/29234)
2. Make `tctl create X` only call `CreateX` methods, and fail if the resource already exists.
3. Create a separate `tctl upsert X` method which only calls `UpsertX` methods (also relates to https://github.com/gravitational/teleport/issues/29234)

However I still think this "quick and dirty" fix to enforce consistency is worthwhile for the time being.

ibeckermayer commented on tool/tctl/common/resource_command.go:
```diff
@@ -856,34 +935,24 @@ func (rc *ResourceCommand) createDevice(ctx context.Context, client auth.ClientI
 		return trace.Wrap(err)
 	}
+	name := dev.Id
+	_, err = client.DevicesClient().GetDevice(ctx, &devicepb.GetDeviceRequest{DeviceId: dev.GetId()})
```
Using multiple means for achieving the same end is unnecessary complexity
```suggestion
	_, err = client.DevicesClient().GetDevice(ctx, &devicepb.GetDeviceRequest{DeviceId: name})
```

ibeckermayer commented on tool/tctl/common/resource_command.go:
```diff
@@ -817,7 +893,13 @@ func (rc *ResourceCommand) createLoginRule(ctx context.Context, client auth.Clie
 	})
-	return trail.FromGRPC(err)
+
+	if err != nil {
+		return trace.AlreadyExists("login rule '%s' already exists, use -f flag to override", rule.Metadata.GetName())
+	}
```
I think `createLoginRule` was already close to fine [as-is](https://github.com/gravitational/teleport/blob/6fdc2fd1df795088466bbe75e48d1fefa1514d62/tool/tctl/common/resource_command.go#L804-L821), presuming that `CreateLoginRule` returns `trace.AlreadyExists` if the login rule already exists. It could just use a `fmt.Printf` at the bottom. (Please manually double check me on that, though).

Regardless, this line doesn't make much sense. That `err` could be anything, not just a `trace.AlreadyExists`.

ibeckermayer commented on tool/tctl/common/resource_command.go:
```diff
@@ -955,7 +1022,8 @@ func (rc *ResourceCommand) createIntegration(ctx context.Context, client auth.Cl
 		return trace.Wrap(err)
 	}
-	fmt.Printf("Integration %q has been created\n", integration.GetName())
+	//fmt.Printf("Integration %q has been created\n", integration.GetName())
```
remove

hegurhurgurk committed:
Apply suggestions from code review

Co-authored-by: Isaiah Becker-Mayer <ibeckermayer@gmail.com>

hegurhurgurk committed:
Update resource_command.go

updating to incorporate recommendations from code review

hegurhurgurk committed:
Update resource_command.go

updated error handling logic in createLoginRule and removed an erroneous comment

ibeckermayer commented on tool/tctl/common/resource_command.go:
```diff
@@ -596,11 +615,18 @@ func (rc *ResourceCommand) createNetworkRestrictions(ctx context.Context, client
+	if err = client.SetNetworkRestrictions(ctx, newNetRestricts); err != nil {
+		return trace.Wrap(err)
+	}
+	fmt.Printf("network restriction '%s' has been %s\n", newNetRestricts.GetName(), UpsertVerb(exists, rc.IsForced()))
```
I still see one remaining instance of `IsForced()`. Let's just delete that method entirely, we don't need to export `rc.force`.

ibeckermayer commented on tool/tctl/common/resource_command.go:
```diff
@@ -521,10 +535,15 @@ func (rc *ResourceCommand) createClusterNetworkingConfig(ctx context.Context, cl
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	exists := (err == nil)
```
Still awaiting an answer on this one.

hegurhurgurk replied:
I agree, this was resolved in the update resource command.go about a week ago. commit f3d21afaa87e071b5da681359e4f22a334472b70

ibeckermayer commented on tool/tctl/common/resource_command.go:
```diff
@@ -808,16 +871,28 @@ func (rc *ResourceCommand) createLoginRule(ctx context.Context, client auth.Clie
 		return trail.FromGRPC(err)
 	}
+	if exists && !rc.force{
+		return trace.AlreadyExists("login rule '%s' already exists, use -f flag to override", rule.Metadata.GetName())
```
```suggestion
		return trace.AlreadyExists("login rule '%s' already exists", rule.Metadata.GetName())
```

telling the user to use `-f` to override in these errors is redundant as the calling code already does that

https://github.com/hegurhurgurk/teleport/blob/c22559cb8839016b87c4b9c7a26a078288178cd2/tool/tctl/common/resource_command.go#L302-L307

(I know you're just copy/pasting what was already there, but let's fix it now). Let's also go through and make all of these `AlreadyExists` errors consistent for all the `CreateHandlers`: start lowercased (except for acronyms), no period, no mention of using -f.

ibeckermayer reviewed (commented):
Nice, looking almost there, just a few more comments in this review, plus please take a look at the unresolved comments from the last review.

Heads up, they may be hidden behind a widget like this:

<img width="877" alt="image" src="https://github.com/gravitational/teleport/assets/13578537/0c4f2bd6-0278-42eb-99ec-4f6eb895ec5a">

hegurhurgurk committed:
updated print statements

also removed the isForced function and changed all instances of isForced() to rc.force

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -356,7 +355,21 @@ func (rc *ResourceCommand) createCertAuthority(ctx context.Context, client auth.
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	fmt.Printf("certificate authority connector %q has been %s\n",
```
```suggestion
	fmt.Printf("certificate authority %q has been %s\n",
```

Copy paste error?

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -407,13 +420,13 @@ func (rc *ResourceCommand) createRole(ctx context.Context, client auth.ClientI,
 		return trace.Wrap(err)
 	}
-	fmt.Printf("role '%s' has been %s\n", roleName, UpsertVerb(roleExists, rc.IsForced()))
+	fmt.Printf("role '%s' has been %s\n", roleName, UpsertVerb(roleExists, rc.force))
```
nit: we're using `%q` in some places, and `'%s'` in other places. One results in `"foo"` and the other `'foo'`. If the point of this PR is to be more consistent, we should fix that :-)

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -521,10 +533,16 @@ func (rc *ResourceCommand) createClusterNetworkingConfig(ctx context.Context, cl
 	}
-
+	var _, err = client.GetClusterMaintenanceConfig(ctx)
+	if err != nil && !trace.IsNotFound(err){
```
```suggestion
```

duplicate line?

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -596,11 +614,18 @@ func (rc *ResourceCommand) createNetworkRestrictions(ctx context.Context, client
 	}
-	fmt.Printf("network restrictions have been updated\n")
+	exists := (err == nil)
+	if !rc.force && exists {
```
This is not a token. Another copy paste error?

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -690,28 +721,60 @@ func (rc *ResourceCommand) createToken(ctx context.Context, client auth.ClientI,
+	}
+	exists := (err == nil)
+	if !rc.force && exists {
+		return trace.AlreadyExists("UI config '%s' already exists", uic.GetName())
```
This doesn't really make sense. There is only one UI config, and it always has the same name. I would just remove the name from the message.

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -729,15 +792,16 @@ func (rc *ResourceCommand) createNode(ctx context.Context, client auth.ClientI,
 	exists := (err == nil)
-	if !rc.IsForced() && exists {
-		return trace.AlreadyExists("node %q with Hostname %q and Addr %q already exists, use --force flag to override",
+	if !rc.force && exists {
```
We should either use `rc.force` everywhere and remove `IsForced` or use `IsForced` consistently.

ibeckermayer replied:
I believe that was done in https://github.com/gravitational/teleport/pull/30473/commits/bf40802d45a16116b346deb2daebfee5e2ce8b0e#diff-0ded7335cb699ba03c82523badbb1b4dd5db4e605d34b89b26b88e3942e1f94dL1469-L1473

zmb3 replied:
Nice

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -756,13 +820,13 @@ func (rc *ResourceCommand) createOIDCConnector(ctx context.Context, client auth.
 		return trace.Wrap(err)
 	}
-	fmt.Printf("authentication connector '%s' has been %s\n", connectorName, UpsertVerb(exists, rc.IsForced()))
+	fmt.Printf("authentication connector '%s' has been %s\n", connectorName, UpsertVerb(exists, rc.force))
```
```suggestion
	fmt.Printf("connector '%s' has been %s\n", connectorName, UpsertVerb(exists, rc.force))
```

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -908,7 +971,7 @@ func (rc *ResourceCommand) createOktaImportRule(ctx context.Context, client auth
 		}
 	}
-	fmt.Printf("Okta import rule '%s' has been %s\n", importRule.GetName(), UpsertVerb(exists, rc.IsForced()))
+	fmt.Printf("okta import rule '%s' has been %s\n", importRule.GetName(), UpsertVerb(exists, rc.force))
```
```suggestion
	fmt.Printf("Okta import rule '%s' has been %s\n", importRule.GetName(), UpsertVerb(exists, rc.force))
```

Okta is the name of a company, it should be capitalized in user-facing messages

Commit:
cleaning up

Commit:
Adds description to CreateHandlers

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -556,7 +555,7 @@ func (rc *ResourceCommand) createClusterMaintenanceConfig(ctx context.Context, c
 	}
 
-	fmt.Printf("maintenance window '%s' has been %s\n", cmc.GetName(), UpsertVerb(exists, rc.force))
+	fmt.Printf("maintenance window %q has been %s\n", cmc.GetName(), UpsertVerb(exists, rc.force))
```
Similar to UI config, there aren't multiple of these.

ibeckermayer replied:
3cffd6fb0160b7e1c313a308ba98992a38ea8e82

Commit:
removes redundant name

hegurhurgurk committed:
Merge branch 'master' into master

zmb3 reviewed (approved):
@ibeckermayer will probably have to buddy this one, but looks good.

hegurhurgurk committed:
Fix for creating integrations

hegurhurgurk committed:
Merge branch 'master' into master

hegurhurgurk committed:
Updated unit test to properly reflect new behaviour

hegurhurgurk committed:
Merge branch 'master' into master

hegurhurgurk committed:
Merge branch 'master' of https://github.com/hegurhurgurk/teleport

ibeckermayer commented:
Buddy PR: https://github.com/gravitational/teleport/pull/33043

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gravitational/teleport?pullRequest=30473) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/gravitational/teleport?pullRequest=30473) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: hegurhurgurk<br/>:x: ibeckermayer<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gravitational/teleport?pullRequest=30473) it.</sub>
